.github/workflows/test-and-push.yml

```diff
@@ -0,0 +1,37 @@
+name: test-and-push,
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - '*'
+env:
+  BRANCH_NAME: ${{ github.head_ref || github.ref_name }}
+jobs:
+  test-and-push:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 22.x
+      - name: Install dependencies
+        run: npm ci
+      - name: Login to GHCR (GitHub Packages)
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build docker image
```

Comment:
we should probably be testing code first before pushing? because then we create bad images

Reply:
This just builds and pushes the image for the SHA tag though, and then we run that for the integration tests. The branch tag only gets pushed once the tests pass, so branch tag should always be good. We've used this pattern elsewhere.